### PR TITLE
Diagnose Circular Typealiases in Protocols

### PR DESCRIPTION
**Pull request by CodaFi**
Fixes a longstanding issue where declarations of the form

``` swift
protocol Foo {
    typealias X = X
}
```

would accidentally crash the compiler by sending the AST visitor into a loop with the typechecker.

The error `type alias 'X' circularly references itself` is now emitted at the declaration site.

### CODE REVIEW

File: include/swift/AST/Decl.h [hunk 1]
```diff
@@ -573,6 +573,9 @@ class alignas(1 << DeclAlignInBits) Decl {
     unsigned : NumTypeDeclBits;
 
     unsigned Recursive : 1;
+
+    /// Whether or not this declaration is currently being type-checked.
+    unsigned BeingTypeChecked : 1;
   };
   enum { NumAssociatedTypeDeclBits = NumTypeDeclBits + 1 };
```
Comment (cwillmor) [inline include/swift/AST/Decl.h]: Needs to be bumped from `+ 1` to `+ 2`.

Comment (CodaFi) [inline include/swift/AST/Decl.h]: I wonder if it's even necessary.  I couldn't find uses for the Recursive flag, but I was worried I would be stepping on toes overwriting it.  Still, it seems like it can be merged.

Comment (CodaFi) [inline include/swift/AST/Decl.h]: @cwillmor Great catch!  Thanks.

**Comment by AlexDenisov** [comment]
Hi @CodaFi, would be better if you could rebase against master. This way your patch will avoid merge-commits, that are not needed usually.

See this article for more details:
https://robots.thoughtbot.com/git-interactive-rebase-squash-amend-rewriting-history

**Comment by CodaFi** [comment]
@AlexDenisov I've been rebasing.  Do you mean I should squash?

**Comment by AlexDenisov** [comment]
@CodaFi I see there are few merge commits ('merge branch master ...'), besides that it could be one commit.
> Do you mean I should squash?

the correct answer might be "rebase and squash" :)

**Comment by CodaFi** [comment]
@AlexDenisov Ah, I see what was happening.  Wrong script!

**Comment by AlexDenisov** [comment]
@CodaFi thank you, not it looks better. You may think about such changes as of one _atomic_ patch ;)

**Comment by swizzlr** [comment]
It seems that the Swift team do not subscribe to the dogmatic view that rewriting history outside mainline is bad!

( :tada: )

**Comment by CodaFi** [comment]
@DougGregor I believe this is ready for review when you're ready!

File: lib/Sema/TypeCheckDecl.cpp [hunk 2]
```diff
@@ -3173,7 +3173,7 @@ class DeclChecker : public DeclVisitor<DeclChecker> {
         TAD->getUnderlyingTypeLoc().setInvalidType(TC.Context);
       } else if (TAD->getDeclContext()->isGenericContext()) {
         TAD->setInterfaceType(
-          TC.getInterfaceTypeFromInternalType(TAD->getDeclContext(),
+        TC.getInterfaceTypeFromInternalType(TAD->getDeclContext(),
                                               TAD->getType()));
```
Comment (DougGregor) [inline lib/Sema/TypeCheckDecl.cpp]: Unnecessary whitespace change here.

Comment (danielphillips) [inline lib/Sema/TypeCheckDecl.cpp]: I would go as far as saying that white space had meaning and now is gone. It was enforcing that getInterfaceTypeFromInternalType is an argument passed to setInterfaceType. 
It's merged now but would you oppose adding this space back in? I'm not one for PR (just to correct this) with such minimal changes but...

**Comment by jtbandes** [comment]
Neat! @CodaFi, do you think this will also fix https://bugs.swift.org/browse/SR-84 ?
